Leia

JimmyS83 opened the pull request:
Latest updates from Master branch picked.

MoojMidge committed:
Initial stream selection re-write

- Not well tested and known to not work correctly

MoojMidge committed:
Remove unused code

MoojMidge committed:
Make normalize function a staticmethod

MoojMidge committed:
Update client priorities and comments

MoojMidge committed:
Fix potential TypeError when sort key is missing

MoojMidge committed:
Short-circuit retries when geo-blocked

MoojMidge committed:
Update stream selections

- Match audio and video containers
- Properly check for IA compatibility
- Properly check for HDR enabled/disabled
- Properly sort HDR enabled streams
- Adjust <Role/> value based on stream selection
- This is not as robust as the old method but is simpler/quicker

MoojMidge committed:
Tidy up

MoojMidge committed:
Remove redundant checks

cas-- committed:
fix: Limit host connections getting subscription feeds

When attempting to view My Subscriptions some users are reporting they
are unable to see any results with following errors in logs:

    ssl.SSLZeroReturnError: TLS/SSL connection has been closed (EOF) (_ssl.c:997)

This is a result on certain systems of attempting to fetch too many
subscription channels (~50 or more) since the current code generates a
thread and connection for every channel feed.

Fixed using a requests session to reuse the connection and setting a
maximum pool size of 5 with blocking enabled to prevent new connections.
There is also a speedup from limiting the pool size (0.5s vs 5s) and the
selected pool size seems a good compromise.

Fixes: https://github.com/anxdpanic/plugin.video.youtube/issues/280

MoojMidge committed:
Update stream selection for multi-language videos

- Fixes default and original audio selection
- Fixes identification/selection of multi-channel audio
- Fixes identification/selection of multi-language audio
- Fixes selection of higher quality audio from non-matching containers
- Fixes edge-cases where selected stream doesn't match label in selection dialog

anxdpanic committed:
Merge pull request #477 from MoojMidge/stream_selection

Update stream selection for multi-language videos

anxdpanic committed:
Merge pull request #478 from cas--/fix/mysubscription-feeds-error

fix: Limit host connections getting subscription feeds

MoojMidge committed:
Update itag format details

MoojMidge committed:
Fix logging error without response

MoojMidge committed:
Store selected client and header details

MoojMidge committed:
Additional request for captions for clients that dont provide them

MoojMidge committed:
Make check for geo-blocked video errors less specific

MoojMidge committed:
Make case of stream title consistent with other titles

MoojMidge committed:
Fix possible regression in capability evaluation

- The meaning of None has changed a few times (#127, #146)
- Add comment to define possible values
- Vorbis audio doesn't seem to be offered by Youtube anymore
- Vorbis appears to have been supported by ISA for a long time

MoojMidge committed:
Enable HLS and adaptive HLS live streams

MoojMidge committed:
Remove disable_dash client property

MoojMidge committed:
Allow HDR when Adaptive MP4 video is selected

MoojMidge committed:
Update server request log messages

MoojMidge committed:
Add default IP param to IP address setting getter

MoojMidge committed:
Add some basic error handling for https session

MoojMidge committed:
Merge branch 'requests'

MoojMidge committed:
Merge branch 'captions'

MoojMidge committed:
Merge branch 'live_streams'

MoojMidge committed:
Implement original single stream live HLS playback

MoojMidge committed:
Merge branch 'live_streams'

MoojMidge committed:
Fix live stream settings options when MPD disabled

MoojMidge committed:
Merge branch 'live_streams'

MoojMidge committed:
Remove Authorisation header from HLS request

- Header is not needed as url has already been retrieved using API

MoojMidge committed:
Fix age gate setting not being applied

MoojMidge committed:
Merge branches 'captions' and 'live_streams'

MoojMidge committed:
Fail earlier if playback is not possible

MoojMidge committed:
Merge branch 'captions'

MoojMidge committed:
Make lang attribute value conform to MPD specs

Also display language in stream quality selector

MoojMidge committed:
Merge branch 'stream_selection'

MoojMidge committed:
Improve handling of dubbed and descriptive audio

Also align MPD manifest with spec and ISA quirks

MoojMidge committed:
Merge branch 'stream_selection'

MoojMidge committed:
Further improve stream labelling

- Use custom video label (Youtube qualityLabel is not accurate)
- Avoid duplicate tokens (from ISA and Youtube descriptions)
- Provide bitrate of audio stream per AdaptationSet (not shown by ISA)

MoojMidge committed:
Merge branch 'stream_selection'

MoojMidge committed:
Remove conditions for 30 fps limit setting

- Should not be linked to HDR setting as there are non-HDR HFR itags
- Should not be disabled if a preferred container is selected

MoojMidge committed:
Improve selection and display of video resolution

- Correctly handles portrait videos
- Correctly handles ultra wide AR videos
- Code is messy due to backwards compatibility with quality settings

MoojMidge committed:
Merge branch 'stream_selection'

MoojMidge committed:
Fix sorting of audio in stream selection dialog

MoojMidge committed:
Update ISA capabilities map and simplify matching

MoojMidge committed:
Improve labelling of more odd video resolutions

MoojMidge committed:
Improve MPD sorting and default quality selection

- This won't fix incorrect ordering by ISA for manual selection
- ISA performs it's own simple sorting that results in H264 being preferred over better AV1 streams

MoojMidge committed:
Merge branch 'stream_selection'

MoojMidge committed:
Standardise/change "dash" in names to "mpd"

MoojMidge committed:
Fix evaluation of single ISA capability

MoojMidge committed:
Improve stream quality and features selection

MoojMidge committed:
Standardise setting attribute order

MoojMidge committed:
Improve MPD manifest language sorting

- ISA will still override this and sort by bandwidth

MoojMidge committed:
Fix incorrect return on create directory failure

MoojMidge committed:
Fix sorting of stream chooser

Sorts based on:
- quality
- preferred language
- original language
- language name (in English...)

MoojMidge committed:
Move get_language_name to kodion context class

MoojMidge committed:
Fix sorting of ISA audio chooser

- Also improve visibility of stream details
- Details in square brackets are max quality level of an adaptive stream
- Streams without square brackets have a single fixed quality level
- Duplication of language in name is unavoidable

MoojMidge committed:
Fix filtering streams when auto selection disabled

MoojMidge committed:
Code tidy up

MoojMidge committed:
Partially fix use history settings

- Split into local and remote history settings

MoojMidge committed:
Fixes for throttling and restricted videos

MoojMidge committed:
Improve client building

MoojMidge committed:
Fix errors in fetching captions

MoojMidge committed:
Dont copy headers if they are not being modified

MoojMidge committed:
Fix json argument of _request redefining import

MoojMidge committed:
Code tidy up

MoojMidge committed:
Allow mulltiple params in call to _set_query_param

MoojMidge committed:
Simplify and tidy up code

MoojMidge committed:
Fix default caption selection

MoojMidge committed:
Simplify update of MPD stream format details

MoojMidge committed:
Split up _generate_mpd_manifest

-

MoojMidge committed:
Use default caption language to label audio track

MoojMidge committed:
Simplify language detection & show ASR indication

MoojMidge committed:
Allow cookies to be used from player page request

weblate committed:
Translated using Weblate (Slovak (sk_sk))

Currently translated at 100.0% (304 of 304 strings)

Translated using Weblate (Catalan (Spain) (ca_es))

Currently translated at 0.9% (3 of 304 strings)

Translated using Weblate (Spanish (Spain) (es_es))

Currently translated at 100.0% (304 of 304 strings)

Translated using Weblate (Slovak (sk_sk))

Currently translated at 99.0% (301 of 304 strings)

Co-authored-by: Christian Gade <gade@kodi.tv>
Co-authored-by: Hosted Weblate <hosted@weblate.org>
Co-authored-by: Jose Riha <jose1711@gmail.com>
Co-authored-by: José Antonio Alvarado <jalvarado0.eses@gmail.com>
Co-authored-by: Patrik Špaňo <patrik.spano@gmail.com>
Translate-URL: https://kodi.weblate.cloud/projects/kodi-add-ons-video/plugin-video-youtube/ca_es/
Translate-URL: https://kodi.weblate.cloud/projects/kodi-add-ons-video/plugin-video-youtube/es_es/
Translate-URL: https://kodi.weblate.cloud/projects/kodi-add-ons-video/plugin-video-youtube/sk_sk/
Translation: Kodi add-ons: video/plugin.video.youtube

anxdpanic committed:
Merge pull request #486 from weblate/weblate-kodi-add-ons-video-plugin-video-youtube

Translations update from Kodi Weblate

anxdpanic committed:
Merge pull request #487 from MoojMidge/stream_selection

Update itag format details

MoojMidge committed:
Make request timeout a parameter of _request

TODO: Make this an advanced setting

MoojMidge committed:
Fix regression with non-logged in playback

MoojMidge committed:
Improve error/debug logging

MoojMidge committed:
Update and tidy up client details

MoojMidge committed:
Use client specific api keys

MoojMidge committed:
Code tidy up

MoojMidge committed:
(Re)Allow tv api key use

- Not enabled by default

MoojMidge committed:
Fix for missing hls manifest url

TODO: combine different stream parsing methods to avoid duplicate code

MoojMidge committed:
Fix different bitrate label vs ISA stream selector

- Also misc related tidy code tidy up

MoojMidge committed:
Fix incorrect level for declaring manifest_url

MoojMidge committed:
Add timeout and error handling for api requests

MoojMidge committed:
Add timeout, fix incorrect args for Session.get

MoojMidge committed:
Fix incorrect httpd request address

MoojMidge committed:
Settings name import tidy up

MoojMidge committed:
Merge pull request #488 from MoojMidge/captions

Subtitles for restricted videos

MoojMidge committed:
Merge branch 'master' into master

MoojMidge committed:
Allow manual make-release workflow trigger

MoojMidge committed:
Update make-release workflow for manual run on forks

MoojMidge committed:
7.0.2

MoojMidge committed:
Add backwords compatibility for Matrix

MoojMidge committed:
Remove unused code

MoojMidge committed:
Update unofficial.patch

MoojMidge committed:
Update make-release workflow formatting and permissions

MoojMidge committed:
Update submit-release workflow for Matrix

MoojMidge committed:
Merge pull request #506 from MoojMidge/master

Client and stream processing updates

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translated using Weblate (French (France) (fr_fr))

Currently translated at 100.0% (304 of 304 strings)

Translated using Weblate (Slovak (sk_sk))

Currently translated at 100.0% (304 of 304 strings)

Co-authored-by: Hosted Weblate <hosted@weblate.org>
Co-authored-by: Jose Riha <jose1711@gmail.com>
Co-authored-by: SlashAndRoses <contact@iandeveseleer.fr>
Translate-URL: https://kodi.weblate.cloud/projects/kodi-add-ons-video/plugin-video-youtube/
Translate-URL: https://kodi.weblate.cloud/projects/kodi-add-ons-video/plugin-video-youtube/fr_fr/
Translate-URL: https://kodi.weblate.cloud/projects/kodi-add-ons-video/plugin-video-youtube/sk_sk/
Translation: Kodi add-ons: video/plugin.video.youtube

MoojMidge committed:
Merge pull request #497 from weblate/weblate-kodi-add-ons-video-plugin-video-youtube

Translations update from Kodi Weblate

MoojMidge committed:
7.0.2.1

MoojMidge committed:
Merge pull request #513 from MoojMidge/master

7.0.2.1

MoojMidge committed:
Fix Matrix ListItemInfoTag class

weblate committed:
Translated using Weblate (Finnish (fi_fi))

Currently translated at 100.0% (329 of 329 strings)

Translated using Weblate (Spanish (Spain) (es_es))

Currently translated at 100.0% (329 of 329 strings)

Translated using Weblate (German (de_de))

Currently translated at 100.0% (329 of 329 strings)

Translated using Weblate (Polish (pl_pl))

Currently translated at 100.0% (329 of 329 strings)

Translated using Weblate (Italian (it_it))

Currently translated at 100.0% (329 of 329 strings)

Translated using Weblate (Spanish (Spain) (es_es))

Currently translated at 100.0% (329 of 329 strings)

Co-authored-by: Alfonso Cachero <alfonso.cachero@gmail.com>
Co-authored-by: Hosted Weblate <hosted@weblate.org>
Co-authored-by: José Antonio Alvarado <jalvarado0.eses@gmail.com>
Co-authored-by: Kai Sommerfeld <ksooo@users.noreply.kodi.weblate.cloud>
Co-authored-by: Marek Adamski <fevbew@wp.pl>
Co-authored-by: Massimo Pissarello <mapi68@gmail.com>
Co-authored-by: Oskari Lavinto <olavinto@protonmail.com>
Translate-URL: https://kodi.weblate.cloud/projects/kodi-add-ons-video/plugin-video-youtube/de_de/
Translate-URL: https://kodi.weblate.cloud/projects/kodi-add-ons-video/plugin-video-youtube/es_es/
Translate-URL: https://kodi.weblate.cloud/projects/kodi-add-ons-video/plugin-video-youtube/fi_fi/
Translate-URL: https://kodi.weblate.cloud/projects/kodi-add-ons-video/plugin-video-youtube/it_it/
Translate-URL: https://kodi.weblate.cloud/projects/kodi-add-ons-video/plugin-video-youtube/pl_pl/
Translation: Kodi add-ons: video/plugin.video.youtube

MoojMidge committed:
Fix error with subtitle selection prompt

MoojMidge committed:
Misc minor fixes

MoojMidge committed:
Updates to allow backports for Kodi 18

MoojMidge commented:
I realised I have added a few bits of code that would not be compatible with Python 2 in the v7.0.2.x releases

Tried to fix this up in https://github.com/MoojMidge/plugin.video.youtube/commits/master/, specifically https://github.com/MoojMidge/plugin.video.youtube/commit/2b87a2de689dfb7c45efc7b2f3cf13a1b42bc351

If you are still interested in updating the Leia branch, would you be able to test these fixes work? If so, I can PR the fixes to the master branch of this repository, so you can in turn update this PR

JimmyS83 commented:
Hello, to be honest I didnt touch isengard branch for a long time, with simple reason, everything just works for me (I didnt experienced some issues, which I can see in master branch at all which have been adressed).

If needed, I can try to pick up recent commits to the Isengard branch (in case they are compatible) - in fact possibly even you can do that, (as I understand you try to write py2 compatible code) and test it in Kodi 18 test installation on PC..   On my "PROD" TV boxes I am following approach "if its not broken, do not touch it" for stability matter :)

MoojMidge commented:
I still have an infrequently used device that is happily running v7.0.1 for movie trailers without errors, and none of my devices were ever consistently impacted by the different causes of the 403 streaming errors, so while the assorted playback issues didn't universally effect everyone, I'd still suspect it's just a matter of time.

I don't have any devices that run Kodi 18, nor the time/motivation to make and test any changes targeting that version on my computer. The changes I referenced previously were just the result of running pylint, after coming across a different bug, and checking for the most obvious Python 2 incompatibilities I may have introduced. No idea if that is everything, or about any of the settings changes that were also made along the way. 

Anyway the code is there if/when you want to have a go at it.

JimmyS83 committed:
updated settings.xml

JimmyS83 committed:
Merge branch 'isengard' into leia

JimmyS83 commented:
> I still have an infrequently used device that is happily running v7.0.1 for movie trailers without errors, and none of my devices were ever consistently impacted by the different causes of the 403 streaming errors, so while the assorted playback issues didn't universally effect everyone, I'd still suspect it's just a matter of time.
> 
> I don't have any devices that run Kodi 18, nor the time/motivation to make and test any changes targeting that version on my computer. The changes I referenced previously were just the result of running pylint, after coming across a different bug, and checking for the most obvious Python 2 incompatibilities I may have introduced. No idea if that is everything, or about any of the settings changes that were also made along the way.
> 
> Anyway the code is there if/when you want to have a go at it.

I tried to cherry-pick all those 100+ commits, and try to apply them as well as keep py2 code in original pieces.. Quite expectedly, there is no easy and smooth flow.. Starting with `TypeError: unsupported operand type(s) for |: 'list' and 'list'` for 
`for key in (item1.keys() | item2.keys()):` 
in `video_info`. 

I am afraid there will be more things to solve to make backward patch working with all theses recent changes.. :(

JimmyS83 committed:
py2 fix, cosmetics

JimmyS83 commented:
@MoojMidge  Now, current status it seems to be at least somehow working (tested just one 1080 video from my subscriptions) - It plays only on 720p with Inputs.Adapt. now, I guess its a toll for som different techniques how to get a stream (and mask client)? When I cherry picked your https://github.com/MoojMidge/plugin.video.youtube/commit/2b87a2de689dfb7c45efc7b2f3cf13a1b42bc351 and add a fix for the .keys(). Would be great, if you would be able to push changes from your repo into into master.. To my surprise that viewkeys things seems to be only missing, at least for very basic test case usage.

JimmyS83 committed:
py2/py3 compatibility

JimmyS83 committed:
py2 encoding fix

JimmyS83 commented:
@MoojMidge I mean Isengard with all these commints from this PR, working on Kodi18, to be more precise :)

MoojMidge commented:
Great work!

Forgot about the changes to dict views. There is another way to do this that saves having to specifically modify the code for Python 2
```
for key in (item1.keys() | item2.keys()):
```
can be changed to:
```
keys = set(item1)
keys.update(item2)
for key in keys:
```

Which is both backwards compatible and also a bit more efficient in Python 2 and Python 3, although it does look uglier. I'll PR that along with the other changes for you to cherry pick over the weekend.

One question though - what is the reason for the removal in this commit:
https://github.com/anxdpanic/plugin.video.youtube/pull/496/commits/312eaaead743286728e756c4eb0d091bd45304c4#diff-9f96c49b6787bd96972525d11f52532a6488c2bc8000c7981e9ef7b2384433a4L67-L70

The reason I ask is because it kinda looks like that code should actually be added back into the master branch. Does it cause issues in Kodi 18?

JimmyS83 commented:
That piece of code was somehow mystery to me also. First I thought, that I missed some of previous commits, try to track it back, but there is SO MUCH commits allong way (even before I pushed that first 6.2x update for Isengard originally, when YT was broken for py2). I even do precautiously re-cherry picked commits in this PR, originally. 

Piece of code popsup because throws error, I believe .use_dash() wasnt there for play_item (which confuses me, because there is commit already before to get rid of dash and replace with mtd, so Ive got suspiction around that piece of code). I compared code to some latest master (and until now I thought it was from master) and this piece of code wasnt there.  I hope I didnt mistaken master and your fork master..  So I remove them and test, if something else is missing or throwing error.

Thats the reason why I reviewed all commits in this PR against master and everything except 2 commits not PR'd to master looks aligned after some rebase and cherry pick work.

Regarding keys, feel free to optimize it yet before pushing to the master, I can later just pick it up to this PR (I dont think it will be merged soon, until these things will be resolved in comments).

MoojMidge commented:
That code was originally made conditional on version in #400, and then removed in #414, however the replacement code was for playback using InputStream.Adaptive in Kodi 20+, while the original code was explicitly for playback using cURL and FFmpeg, and therfore not mutually exclusive. That's why I thought it may actually need to be added back in to the master branch.

The reason the original code would now cause an error is because `.use_dash()` was renamed to `.use_dash_video()` in [`6264b40`](https://github.com/anxdpanic/plugin.video.youtube/commit/6264b408f7c34c655b8a70b4c0617af9fb1710a6#diff-6941c194620aa86efcfdaea452fa8a63ac9c78489548db984c8d81c258fb12f7), and finally to `.use_mpd_video()` in [`2ba69cc`](https://github.com/anxdpanic/plugin.video.youtube/commit/2ba69cce1a8f9d904dbaf1f5abad45e7207602f4#diff-6941c194620aa86efcfdaea452fa8a63ac9c78489548db984c8d81c258fb12f7).

You can try renaming it and seeing whether there are any issues with video playback if you disable MPEG-DASH. I'll try the same later next week.

JimmyS83 committed:
cosmetics @comtaibility

JimmyS83 commented:
```
if not play_item.use_mpd_video() and not settings.is_support_alternative_player_enabled() and \
            play_item.get_headers() and play_item.get_uri().startswith('http') and major_version < 20:
        play_item.set_uri('|'.join([play_item.get_uri(), play_item.get_headers()]))
```

works in sense, no compile error, video play both with and without mpeg dash. But the same results Ive got without this code also @Kodi18.

JimmyS83 commented:
Btw., I noticed, that with older code I get 1080p video, even 4k if its available on YT, but with all these new commits I get max. 720p  (Inputstream Adapter without limitation used).

MoojMidge commented:
Can you get a debug log of this happening?

One of the main changes in all these commits was unlinking stream type and stream quality, but some of those changes rely on InputStream.Adaptive functionality that may not be available in Kodi 18

JimmyS83 commented:
Sure, here is it. Dont mind addon version, its not changing as I fiddling with code.
Example video: https://www.youtube.com/watch?v=DuC3p0_Kh-o
Old code, which returns 2160p 60fps - https://pastebin.com/013DB17w
New code, which returns 720p - https://pastebin.com/nSPeY91X

MoojMidge commented:
One of the other main changes in all these commits was to remove almost all the useful debug logging...

I can't access the old code log (it's not important), but the new code log does not appear to show that any of the adaptive streams are being processed, instead the stream list is only being populated with the non-adaptive formats, that are limited to 720p.

There aren't too many reasons why this would happen, either the HTTP server is not running or MPEG-DASH is not enabled. Maybe it was left disabled after previous testing?

To check the HTTP server is running, if you disable and re-enable the Youtube video plugin, you should see the following few log messages, or something similar, in your debug log:
```
debug <general>: [plugin.video.youtube] send_notification: |check_settings| -> |{"use_httpd": true, "httpd_port": 50152, "whitelist": "", "httpd_address": "0.0.0.0"}|
debug <general>: [plugin.video.youtube] YouTube service initialization...
debug <general>: [plugin.video.youtube] HTTPServer: Starting |0.0.0.0:50152|
debug <general>: [plugin.video.youtube] HTTPServer: Serving on |0.0.0.0:50152|
```

Assuming the HTTP server is running and MPEG-DASH is enabled, then there may be a mismatch in your existing stored settings values and the new modified settings definition file.

Someone else reported a similar issue, but it shouldn't persist after changing any of the settings values. However Kodi 18 may handle writing the settings values differently to Kodi 19+ (I can't remember anymore).

You could try the following to double check all the relevant settings are correct:

1. Open Youtube plugin settings > MPEG-DASH
2. Enable `Use MPEG-DASH`
3. Enable `Use for videos`
4. Set `Maximum video quality` to whatever resolution you want playback to be limited to
5. Open `Stream features` and select everything, or unselect any formats/features you specifically don't want to use, then press `OK`
6. Set `Stream selection` to `Automatic` or `Automatic + Quality selection` [1]
7. Set `Use for live streams` to `Adaptive HLS` to use InputStream.Adaptive to play HLS Live Streams [2]
8. Press `OK` in the settings dialog to save the updated settings values.
9. Try video playback again.

Notes:
[1] The `Automatic` may not work in Kodi 18, so `Quality selection` may need to be set as the default for this branch. You can check whether the different settings values work by playing a video then opening `Video OSD > Settings > Video settings > Video stream`.

`Automatic` should add a video stream with all available qualities combined into one, that InputStream.Adaptive can automatically choose from, depending on display resolution, max bandwidth settings, and actual available bandwidth. The stream selection based on available bandwidth is only going to work in Kodi 20+, but I am not sure whether the other stream selection logic will work in Kodi 18.

`Quality selection` should add individual video streams for each available video quility, so that it can be manually selected on-the-fly during video playback.

[2] `MPEG-DASH` live streams won't work currently. `HLS` uses cURL + ffmpeg to play a live stream using the best available non-adaptive format and is slow to load, `Multi-stream HLS` uses cURL + ffmpeg to allow all available formats to be selected but can be very slow to load. `Adaptive HLS` is the same as `Multi-stream HLS` but uses InputStream.Adaptive and should be much faster.

JimmyS83 commented:
Unfortunatelly no luck here. I keep live stream aside, as even basic videos are affected by this. I deleted usedata, so settings was recreated again with new code. 2/3 i always there. I tried all three Stream Selection options. I setup Input.Adaptive from Autom-Manual choose.. Considering I never get question from it, it seems its not used as you wrote, in all three Stream Selection modes :( HTTP Server inicalization is present, wouldnt expect problems there...

Vanilla settings exmaple: https://pastebin.com/PhHBkx3s
Debug log: https://pastebin.com/vg032F16

MoojMidge commented on resources/lib/youtube_plugin/youtube/helper/video_info.py:
```diff
@@ -1713,3 +2178,8 @@ def filter_qualities(stream_data, container, sorted_qualities, fps_limit):
+            _settings.httpd_listen(for_request=True),
+            _settings.httpd_port(),
+            self.video_id
+        ), main_stream
```
This last return statement is never reached so the correct url of the generated manifest is never used as a stream option. This manifest contains all the adaptive stream formats, which is one of the reasons why only the standard format (max 720p) streams are being played.

The "success" return in the `try/except` block  should be updated as `ipaddress` and `stream_info` are not defined anywhere. This may be from an older version of the code.

Something like the following, adapted from current master, should work:
```
        filepath = '{0}{1}.mpd'.format(basepath, self.video_id)
        success = None

        try:
            mpd_file = xbmcvfs.File(filepath, 'w')
        except (IOError, OSError):
            self._context.log_error('Unable to open MPD file path for writing')
        else:
            try:
                success = mpd_file.write(str(out.encode('utf-8')))
            finally:
                mpd_file.close()

        if not success:
            return None, None
        return 'http://{0}:{1}/{2}.mpd'.format(
            _settings.httpd_listen(for_request=True),
            _settings.httpd_port(),
            self.video_id
        ), main_stream
```

MoojMidge commented on resources/lib/youtube_plugin/kodion/impl/abstract_settings.py:
```diff
@@ -193,62 +209,60 @@ def get_location(self):
             return []
+        selected = self.get_int(SETTINGS.MPD_QUALITY_SELECTION, 4)
+        return [quality for key, quality in self._QUALITY_SELECTIONS.items()
+                if selected >= key]
```
dict iteration order matching insertion order is a Python 3.7+ feature. The order of the quality selections is important when sorting and selecting the available adaptive streams - if it is not in order then the quality comparisons may fail and the streams will be skipped.

This may fail in Kodi 19+ as well depending on the version of Python used, so I will update this in master to be:
```
    return [quality
                for key, quality in sorted(
                    self._QUALITY_SELECTIONS.items(),
                    reverse=True
                )
                if selected >= key]
```

For Kodi 18 you can additionally change `self._QUALITY_SELECTIONS.items()` to be `self._QUALITY_SELECTIONS.viewitems()`

MoojMidge commented on resources/lib/youtube_plugin/kodion/impl/xbmc/xbmc_items.py:
```diff
@@ -47,31 +64,53 @@ def to_play_item(context, play_item):
         list_item.setContentLookup(False)
         list_item.setMimeType('application/xml+dash')
-        list_item.setProperty(inputstream_property, 'inputstream.adaptive')
+        list_item.setProperty('inputstream', 'inputstream.adaptive')
```
It was removed in line 59-63, but the property name needs to be 'inputstreamaddon' for Kodi 18.

MoojMidge commented on resources/lib/youtube_plugin/kodion/impl/xbmc/xbmc_items.py:
```diff
@@ -47,31 +64,53 @@ def to_play_item(context, play_item):
+
+        list_item.setContentLookup(False)
+        list_item.setMimeType(mime_type)
+        list_item.setProperty('inputstream', 'inputstream.adaptive')
```
The property name needs to be 'inputstreamaddon' for Kodi 18.

MoojMidge reviewed (commented):
I think there are a few problems with the cherry picking.

I have left some specific code comments based on a file compare, but I don't know if that will be sufficient to fix the video playback of the adaptive streams.

JimmyS83 commented:
Thanks for taking time to look on this.. I will try to test picked details later on, if these helps K18 with adaptive.. Hopefully it will work and we could focus how to implement changes.

 I think we need to agree about handling these differences.. I am glad, that you will try to make some more backwards compatible code in master python wise, but what about Kodi version specific mentioned? Should we change it to K18 in Isengard only "hard coded" way?   I mean, decide, what I should change in this PR for K18, and what its possible to leave solved in master in future..

JimmyS83 committed:
Inputstream and K18 compat fix

JimmyS83 commented:
I just tested, and it works much better now. Inpustream adaptive seems to been used. 

Maybe would be possible to somehow use `context.get_system_version().get_version()` from context also for `abstract_settings`, and include both `.items()` and `.viewitems()` in master code, to keep things more simple.. In terms of tracking changes ? I know that this branch is dedicated to K18 only, and master is somehow only for Nexus or smth like this... But considering backporting just once for a while... 

Only minor thing is, that with vanilla settings, lowest quality is choosed by default when _Manual video selection_@Inputstream adaptive choosed 

![2023-10-16 15_13_44-Kodi](https://github.com/anxdpanic/plugin.video.youtube/assets/7437057/607fdd51-380e-4e44-9aa2-f01e19b1d5d1)

and SD selection choosed when _Automatic selection_@Inputstream adaptive choosed 
![2023-10-16 15_28_35-Kodi](https://github.com/anxdpanic/plugin.video.youtube/assets/7437057/dae37186-26ca-4d19-914b-9ddef43577ee)

and nothing happens (circle spinning) when _Manual select all streams_@Inputstream adaptive choosed..

Max resolution settings are to Max@Inputstream
![2023-10-16 15_33_34-Kodi](https://github.com/anxdpanic/plugin.video.youtube/assets/7437057/a4d6bebb-5f46-4088-a38b-cbe599dbb8d2)

MoojMidge commented:
Prior to Kodi 20, the stream selection options in InputStream.Adaptive were kinda useless, and even now, only the Automatic/Adaptive setting is actually useful for general purpose viewing IMO, with the other settings only of use for testing.

For Kodi 18, you should just stick to Automatic. The reason it is choosing a lower resolution would be dependent on your setup and settings. I would take a guess that you are running Kodi in a window and InputStream.Adaptive is choosing the closest stream resolution that matches your window size. If you select `Ignore Display Resolution` it should try to pick the best quality stream instead.

The purpose of the stream selection options in the Youtube addon is to workaround the limitations of the InputStream.Adaptive stream selection logic, by providing an automatic option (that is also adaptive in Kodi 20+), but also to have properly filtered and sorted manual quality selections, because the manual selection options don't work too well in InputStream.Adaptive.

MoojMidge commented:
> I think we need to agree about handling these differences.. I am glad, that you will try to make some more backwards compatible code in master python wise, but what about Kodi version specific mentioned? Should we change it to K18 in Isengard only "hard coded" way? I mean, decide, what I should change in this PR for K18, and what its possible to leave solved in master in future..


> Maybe would be possible to somehow use `context.get_system_version().get_version()` from context also for `abstract_settings`, and include both `.items()` and `.viewitems()` in master code, to keep things more simple.. In terms of tracking changes ? I know that this branch is dedicated to K18 only, and master is somehow only for Nexus or smth like this... But considering backporting just once for a while...

I'm not the right person to ask, that would be up to @anxdpanic

My 2 cents, FWIW, is that there is not much interest in the isengard branch, so whatever way is easier for contributions to be made would be best.

Personally, I would forget about the isengard branch and create a new leia branch instead. It should be based on master, with specific changes to support Python 2, rather than trying to add conditional statements for the Python 2 changes.

The problem with the conditional statements is that it changes the whitespace, indentation and structure of the code, making the diffs harder to read and therefore making it more difficult to see where the code has changed.

If the code was simply changed to use `.view(keys|values|items)()`, add `.encode('utf-8')` and `.decode('utf-8')`, and use other Python 2 specific changes where required, with a block of code at the beginning of the module to deal with differences in imports and module names, then it would be much easier to rebase the branch when the master branch changed, rather than having to cherry pick changes manually.

JimmyS83 committed:
Inputstream adaptive - Manual sorting

When using Inpustream adaptive with "Manual select streams", by default it choose first resolution. Lets make this higher one.

JimmyS83 commented:
> If you select `Ignore Display Resolution` it should try to pick the best quality stream instead.

Your guess is right. Automatic works as you wrote, so I guess the user experience on K18 could be the same as with master with this #PR.



> Personally, I would forget about the isengard branch and create a new leia branch instead. It should be based on master, with specific changes to support Python 2, rather than trying to add conditional statements for the Python 2 changes.

I think code from Isengard with this commit would be very close to this. I mean regarding keys/encode/decode, six imports, all should be there and these all cherrypicks I believe forward code to the master as much possible. One advantage of conditionals is, that when merging master code after 6months, you can see which previously py2 relevant code is touched/changed in master. But yes, its up to @anxdpanic. My initial reason to do all this is still the same, try to keep older versions of Kodi with YT availability, like away from big breaks.

MoojMidge commented:
> Your guess is right. Automatic works as you wrote, so I guess the user experience on K18 could be the same as with master with this #PR.

Almost. In the master branch there is this code that forces InputStream.Adaptive into automatic stream selection mode if that is selected in the Youtube addon settings, and in Kodi 20+ the automatic stream selection is adaptive so it will automatically change streams based on network conditions.
https://github.com/anxdpanic/plugin.video.youtube/blob/10fecdecd4e6507cfc50d0b1e9cc468aed2ada88/resources/lib/youtube_plugin/kodion/impl/xbmc/xbmc_items.py#L72-L73

In Kodi 18, there appears to be a lot of information missing from the stream selection dialog. The first stream is meant to be labelled as "Automatic", with the other streams labelled based on the fixed video resolutions they offer.

The first stream essentially combines all the other streams into one, and is the only stream that will allow the automatic stream selection in InputStream.Adaptive to work properly i.e. InputStream.Adaptive won't automatically switch from stream 1 (Automatic) to stream 2 (1080p as an example), but it will switch between the sub-streams in stream 1 to change the resolution if automatic stream selection mode is enabled.

That is why automatic is normally the best option, but the Youtube addon adds the fixed resolution streams if you still want to manually change the selected stream, while still keeping the InputStream.Adaptive stream selection mode in automatic.


> I think code from Isengard with this commit would be very close to this.

It should be, yes, but there appears to be some legacy code that tries to work in Python versions older than v2.7 and also v3+.

As an example there is stuff like `list(some_dict.keys())` in the code. This is also in the master branch, but for a codebase targetting Python 3 this should be `some_dict` or `some_dict.keys()` or `list(some_dict)` depending on the usage, and a Python 2.7 version would then only need to change `some_dict.keys()` to `some_dict.viewkeys()`.

While the legacy code is unnecessarily slow, it takes time and effor to update, and as the code still works, it is not a big priority.  However Python 2.7 was only used with Krypton and Leia, so those changes can't be made if the isengard branch is still targetting Kodi versions that are two major versions behind Krypton.


> I mean regarding keys/encode/decode, six imports, all should be there and these all cherrypicks I believe forward code to the master as much possible. One advantage of conditionals is, that when merging master code after 6months, you can see which previously py2 relevant code is touched/changed in master.

Cherry picking seems like a lot of effort to me, and I guess it varies from person to person, but to me the whitespace and indentation changes created by adding the conditionals makes it more difficult to see what has changed, in terms of actual logic/functionality changes, when doing a git diff.

This then makes it more difficult to rebase (or cherry pick changes) when the master branch changes, as you need a more thorough understanding of code that someone else has changed, to make sure that any changes to the master code match the identation levels of the older branch.

I think https://github.com/anxdpanic/plugin.video.youtube/pull/496#discussion_r1359864288 is an example of this.


> But yes, its up to @anxdpanic. My initial reason to do all this is still the same, try to keep older versions of Kodi with YT availability, like away from big breaks.

Your previous PR was merged, so I don't think there is going to be any issues in you continuing to update the branch as you see fit.

JimmyS83 commented:
I agree. But creating Leia compatible version on top of clean master isnt something, which I could pull off, quickly or without errors :)  To be honest, doin a "fix" isengard branch from total disfunctional previously was made just beacuse Leia branch missing here. And hopefully it worked for some users, which stays on a few <Leia versions for whatever reason. In the real world, I can test (at least a bit, not much frequent YT user) just Leia version, not the latest ones..

How about suggest anxdpanic create Leia, and base it on the code from this PR - its kinda mix of latest master bits, but also conditional code. Having scope Leia only, then we could get rid of conditional code, and in the same time try to follow master as much as possible. Writing we, because you are more experienced, also maintaining master, so you will spot a lot more things needs to be aligned than me (not only because you are working with them and writing master)..

JimmyS83 commented:
Any updates? @anxdpanic @MoojMidge ?  Using this version on Kodi18 without problems so far (didnt check possible new commits on master branch till this big cherry picking...)

MoojMidge commented:
Have actually removed some of the Python 2 specific code in a new PR.

It seems counter-intuitive, but I hope that will actually allow for easier backporting, by limiting the number of changes required in the main branch, and instead limit changes in older branches to imports (i.e. Kodi Six and __future\__).

If I get time will see how that goes, otherwise can merge this as-is.

JimmyS83 commented:
Oh boy, there are tens of commits in #543.. Would you please then sent me a bit more detailed info, what would break, when it gets merged? I am not sure if I would have time to dig deep to all these changes :(

_otherwise can merge this as-is_
To be honest, as I am using almost 100% time only Kodi 18, the best would be take this and create branch Leia. Its propably running fine on isengard, but I didnt test it for that version of Kodi, and guys depending on isengard could tak version from leia branch.. It makes more sense to me.. Since K18 was last version with py2, I think it makes sense trying to keep that version not-so-obsolete from all pre-py3 versions of Kodi.

MoojMidge commented:
One of the aims of #542 was to allow the plugin to mostly work in Kodi 18, with some minor compatibility shims in place. This could be maintained without necessarily needing to know what the ongoing changes actually are.

I'll see if this actually pans out in practice


> create branch Leia

Probably, but it is all more work that people don't have time for. Even doing this much to support Kodi 18 is going beyond reasonable expections IMO, but here we are anyway. Not going to stress too much about even older versions.

MoojMidge commented:
Yep, it seems to work without having to change too much. Only main issue is that `RunScript` and `RunPlugin` don't work when using the new settings format.

I'll update #542 to include the necessary changes in the main branch.

Then all that needs to be done for Leia is to rebase and add some import shims, but it is basically the exact same code as the main branch

MoojMidge commented:
#542 has been updated
[MoojMidge/plugin.video.youtube/wip-leia](https://github.com/MoojMidge/plugin.video.youtube/tree/wip-leia) contains all that is required for it to work in Leia i.e. just change the `addon.xml`

Suggest that a Leia specific branch, apart from changing `addon.xml`, should also add some new main directory listing items that include the maintenance and user management actions that don't work through the settings dialog, but apart from that everything seems to work with the same code.

rsantellan commented:
@MoojMidge I installed from zip the branch to work on libreelec 9.2.6 that has kodi 18.9 (Is an old raspberry pi 1 B+) and on the logs I have this
```
2024-03-14 18:00:58.055 T:1800057728   ERROR: [plugin.video.youtube] Player response failed for video_id: LubR2sFpT_E, using android client (logged out)
                                            Traceback (most recent call last):
                                              File "/storage/.kodi/addons/plugin.video.youtube/resources/lib/youtube_plugin/youtube/helper/video_info.py", line 944, in _request
                                                if signature:
                                              File "/storage/.kodi/addons/script.module.requests/lib/requests/models.py", line 960, in raise_for_status
                                                raise HTTPError(http_error_msg, response=self)
                                            HTTPError: 400 Client Error: Bad Request for url: https://www.youtube.com/youtubei/v1/player?prettyPrint=false&key=AIzaSyA8eiZmM1FaDVjRy-df2KTyQ_vz_yYM39w
2024-03-14 18:00:58.058 T:1800057728   ERROR: Traceback (most recent call last):
2024-03-14 18:00:58.059 T:1800057728   ERROR:   File "/storage/.kodi/addons/plugin.video.youtube/resources/lib/youtube_plugin/youtube/helper/yt_play.py", line 50, in play_video
2024-03-14 18:00:58.060 T:1800057728   ERROR:     ui.show_notification(message=e.get_message())
2024-03-14 18:00:58.062 T:1800057728   ERROR:   File "/storage/.kodi/addons/plugin.video.youtube/resources/lib/youtube_plugin/youtube/client/youtube.py", line 103, in get_video_streams
2024-03-14 18:00:58.064 T:1800057728   ERROR:     'ns': 'yt',
2024-03-14 18:00:58.066 T:1800057728   ERROR:   File "/storage/.kodi/addons/plugin.video.youtube/resources/lib/youtube_plugin/youtube/helper/video_info.py", line 896, in load_stream_infos
2024-03-14 18:00:58.067 T:1800057728   ERROR:   File "/storage/.kodi/addons/plugin.video.youtube/resources/lib/youtube_plugin/youtube/helper/video_info.py", line 1348, in _get_video_info
2024-03-14 18:00:58.068 T:1800057728   ERROR:     'vp9': 0.75,
2024-03-14 18:00:58.069 T:1800057728   ERROR:   File "/storage/.kodi/addons/plugin.video.youtube/resources/lib/youtube_plugin/youtube/helper/video_info.py", line 952, in _request
2024-03-14 18:00:58.070 T:1800057728   ERROR: YouTubeException: Player response failed for video_id: LubR2sFpT_E, using android client (logged out)
2024-03-14 18:00:58.071 T:1800057728   ERROR: [plugin.video.youtube] None
2024-03-14 18:00:58.149 T:1938042896   ERROR: Playlist Player: skipping unplayable item: 0, path [plugin://plugin.video.youtube/play/?video_id=LubR2sFpT_E]
```

I can't update to latest version because of raspberry pi limitations, any idea how to fix this ?

MoojMidge commented:
I do not have the capacity to provide any support for this.

This branch is not up to date and won't work.

As indicated in the post directly above yours, the current master branch will work in Kodi v18 if you update the `addon.xml`.

MoojMidge commented:
@rsantellan - https://github.com/anxdpanic/plugin.video.youtube/issues/623#issuecomment-2001886194

rsantellan commented:
> I do not have the capacity to provide any support for this.
> 
> This branch is not up to date and won't work.
> 
> As indicated in the post directly above yours, the current master branch will work in Kodi v18 if you update the `addon.xml`.

Hi @MoojMidge just one question, this is the line I have to change ? 

https://github.com/anxdpanic/plugin.video.youtube/blob/8d6bb622a11826244ee2048a91db327aeee6fe32/addon.xml#L4

Regards and thanks!

MoojMidge commented:
You don't need to change anything now, the release I linked you to already includes the required changes. Just download one of the zip files with `leia` in the name.

rsantellan commented:
> You don't need to change anything now, the release I linked you to already includes the required changes. Just download one of the zip files with `leia` in the name.

Hi I downloaded the branch an that is the error that I posted. I will to debug it and compare with master that is working on another raspberry pi.

MoojMidge commented:
This is not the correct place for this discussion.

The error you posted does not occur when using an up-to-date branch.

The modification you are asking about is not required when using the release you were directed, which is based on the current master branch. No need to debug or compare with anything.

Please read the link you were directed to and just download and install the correct zip file. Create a new issue and post full details if you are still having an issue.

MoojMidge commented:
@JimmyS83 - I am going to close this as it is not required anymore.

Only issue with using the master branch is that some user and maintenance actions can only be accessed through the settings, which won't work in Kodi 18.
